File: epi_judge_python/is_list_palindromic.py
from list_node import ListNode
from test_framework import generic_test


def is_linked_list_a_palindrome(L: ListNode) -> bool:
    if not L or not L.next:
        return True
    sl = fa = L
    while fa and fa.next and fa.next.next:
        sl = sl.next
        fa = fa.next.next
    sec = sl.next
    rev = L
    t = rev.next
    rev.next = None
    while rev is not sl:
        t2 = t.next
        t.next = rev
        rev = t
        t = t2
    if not fa.next:
        sl = sl.next
    while sl and sec:
        if sec.data != sl.data:
            return False
        sl = sl.next
        sec = sec.next
    return True


if __name__ == "__main__":
    exit(
        generic_test.generic_test_main(
            "is_list_palindromic.py",
            "is_list_palindromic.tsv",
            is_linked_list_a_palindrome,
        )
    )

File: epi_judge_cpp/is_list_palindromic.cc
#include "list_node.h"
#include "test_framework/generic_test.h"

shared_ptr<ListNode<int>> Reversed(shared_ptr<ListNode<int>> st,
                                   shared_ptr<ListNode<int>> en) {
  if (st != en) {
    auto left{st}, middle{st->next}, right{st->next->next};
    while (middle != en) {
      middle->next = left;
      left = middle;
      middle = right;
      right = right->next;
    }
    middle->next = left;
  }
  st->next.reset();
  return en;
}

bool Palindrome(shared_ptr<ListNode<int>> left,
                shared_ptr<ListNode<int>> right) {
  while (left) {
    if (left->data != right->data) {
      return false;
    }
    left = left->next;
    right = right->next;
  }
  return true;
}

bool IsLinkedListAPalindrome(shared_ptr<ListNode<int>> L) {
  int count = 0;
  for (auto ptr{L}; ptr; ptr = ptr->next) {
    count++;
  }
  if (auto half = count / 2) {
    auto en{L};
    while (--half) {
      en = en->next;
    }
    auto right = count % 2 == 0 ? en->next : en->next->next;
    auto left = Reversed(L, en);

    return Palindrome(left, right);
  }
  return true;
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"L"};
  return GenericTestMain(args, "is_list_palindromic.cc",
                         "is_list_palindromic.tsv", &IsLinkedListAPalindrome,
                         DefaultComparator{}, param_names);
}
